Bootstrap Sphinx documentation from MetalK8s 1.X

gdemonet opened the pull request:
**Component**: docs
**Type**: feature

**Context**:

- Complex features, like cluster expansion and upgrade, are being implemented without being documented properly, and newcomers (or ourselves if returning from a break) are unaware of how to use them
    - **Documentation effort is _necessary_**

- The MetalK8s 1.X project is currently documented at ReadTheDocs; we can use the same approach for now (and reuse the code)

**Summary**:

- Copy code from MetalK8s 1.X to setup Sphinx RST documentation tree

- Add `tox` environments:
    - `pip-compile`: to generate `requirements.txt` files from `.in` files (static declaration of targets in `tox.ini`)
    - `docs`: to build documentation (basically interface with the `make` build system in `docs/`)

- Bootstrap _Developer Guide_ (see #1069) from existing documents
    - _Architecture Documents_ contains design documents that are useful for developers to know about and refer to when implementing or maintaining features
    - _Best Practices_ is a guide for developers to follow when proposing changes to the project, and should be updated frequently to reflect the project standards

- Build documentation as part of the CI
    - PDF output is saved to Eve artifacts

**Acceptance criteria**:

- Generated documentation is available as a result of a CI build (and successful documentation build is a blocking condition)

- As a developer/technical writer, I can run simple `tox` commands to generate docs from local sources
    - `tox -e docs` builds the HTML static files into `docs/_build`
    - `tox -e docs -- latexpdf` builds the PDF documentation (`tox -e docs -- html latexpdf` builds both)
    - `tox -e docs -- livehtml` watches the `docs/` sources and builds every change, making the docs available at `localhost:8000`

- I can avoid installing dependencies using Docker to generate the docs
    - `docker build -t docs-builder:latest -f docs/Dockerfile .` to build the Docker image
    - `docker run --rm -v "$(pwd)":/usr/src/metalk8s docs-builder:latest <build_cmd>` to use the build container with `<build_cmd>` being `html`, `latexpdf`, ... or `help` to see a list of valid `<build_cmd>`s

- Issues #951, #952, #953, #1050, #1069, #1070, #1071, #1075 can actually be implemented and reviewed without reading through RST files

gdemonet committed:
Add tox env to compile Python requirements

Issue: GH-951

gdemonet committed:
Generate required logos

Both the HTML and PDF docs require new logos to be generated.

Issue: GH-951

gdemonet committed:
Rename 'diagrams' target to 'check-uml'

Since diagrams built this way are not required by the docs building, the
target name should reflect its expected usage.

Issue: GH-951

gdemonet committed:
Add help for additional targets

Issue: GH-951

gdemonet committed:
Add tox env to interface with docs/Makefile

The "build.cmd" file is used to abstract Windows and Unix differences
when building the docs/ Makefile or make.bat targets.

Issue: GH-951

gdemonet committed:
Initialize developer documentation

We had a few documents worth gathering in a "Developer Guide" (as should
be introduced by GH-1069), which we organize into two sections: one for
architecture documents, the other for development best practices.

Issue: GH-1073

gdemonet committed:
Build docs in CI

We introduce a Dockerfile for building docs as well, which makes sense
to extract into the docs/ directory, since generating PDFs requires a
lot of dependencies that developers/writers may not want to install in
their environment.

Issue: GH-1073

gdemonet committed:
Upload documentation to artifacts

Issue: GH-1073

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1073-bootstrap-sphinx-docs` into
`development/2.0`
* [w/2.1/feature/1073-bootstrap-sphinx-docs](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/1073-bootstrap-sphinx-docs) will be merged into `development/2.1` 
* [w/2.2/feature/1073-bootstrap-sphinx-docs](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/1073-bootstrap-sphinx-docs) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf

NicolasT commented on docs/Dockerfile:
```diff
@@ -0,0 +1,42 @@
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/*
+
+RUN python3.6 -m pip install tox && \
```
Why can't we use `tox` from packages?

gdemonet replied:
It didn't ship a recent enough version... 2.5 in the package, we need 3.4 to use `commands_pre` in the `[testenv:tests]` section. Does it warrant a comment in the Dockerfile?

NicolasT replied:
Yes.

NicolasT commented on docs/Makefile:
```diff
@@ -1,16 +1,50 @@
+	@echo $(call _helpline,check-uml,to generate PNG/SVG from UML sources locally)
+	@echo $(call _helpline,livehtml,to automatically rebuild HTML files)
+
+.PHONY: help Makefile
```
Why `Makefile`? This is part of the 1.x one as well, though not sure/clear why :P

gdemonet replied:
It's in the generated `Makefile` from `sphinx-quickstart`, so I guess it makes sense to leave it.

I assumed it's used for the catch-all target defined below, which checks `Makefile` as a requirement (not sure why either):

```
%: Makefile
    @$(SPHINXBUILD) ...
```

NicolasT replied:
OK if it's part of the generated `Makefile`, and your rationale makes sense.

This is a good reason to have a 'plain import of fresh Sphinx quickstart project' commit as the base for all of this...

gdemonet replied:
I'll restart with a fresh `sphinx-quickstart`

NicolasT commented on docs/developer/index.rst:
```diff
@@ -0,0 +1,7 @@
+
+.. toctree::
+
+   architecture/index
```
I don't think you have to specify `index`?

gdemonet replied:
I'm getting warnings if I remove them:

```
.../metalk8s/docs/developer/index.rst:4:toctree contains reference to nonexisting document 'developer/architecture'
```

NicolasT replied:
Oh, my bad then.

NicolasT commented on eve/main.yml:
```diff
@@ -76,6 +77,29 @@ stages:
+          name: Upload PDF documentation
+          source: docs/_build
+          urls:
+            - ['docs/MetalK8s.pdf', 'latex/MetalK8s.pdf']
```
Will this also upload HTML renderings?

gdemonet replied:
Oh I forgot I wanted to dive deeper into this part... So the syntax used above should mean "Retrieve `latex/MetalK8s.pdf` from the build directory `docs/_build` and upload it to artifacts under `docs/MetalK8s.pdf`", according to Eve docs ( which I can't link to --" ). However, it seems it's not renaming anything, so we only get a `latex/MetalK8s.pdf` in the artifacts directory.

Now, your point was about having the HTML renderings - should I upload everything under `docs/_build` then?

NicolasT replied:
Yeah, just do all of `docs/_build` into a `docs/` target?

NicolasT commented on tox.ini:
```diff
@@ -9,6 +9,52 @@ skip_install = true
+setenv =
+    CUSTOM_COMPILE_COMMAND = tox -e pip-compile
+commands =
+    bash -c 'pip-compile \
```
Why is this wrapped in `bash`?

gdemonet replied:
So we can call it on both MacOS and Linux to generate different `requirements-$(uname -s).txt` files (since their requirements differ).

NicolasT replied:
Argh. K. It'd be really great to find a way to remove this split between Linux and OSX `requirements`...

NicolasT reviewed (changes requested):
Couple of questions/remarks.

What you may want to do is to create a fresh Sphinx 'quickstart' project, overwriting `docs/Makefile`, as a very first commit, then continue with the commits you have. That way it's very clear which changes we made compared to a 'standard' Sphinx project.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @alexandre-allard-scality

* @gdemonet

* @slaperche-scality

* @ramanduh

* @ballot-scality

* @teddyandrieux

* @tcarmet

* @lucieleonard

* @zempashi

* @sayf-eddine-scality

* @carlito-scality

* @jbertran

* @thomasdanan

* @monpote

* @nootal

* @nicolast

* @alxf





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

gdemonet commented:
> create a fresh Sphinx 'quickstart' project, overwriting docs/Makefile, as a very first commit, then continue with the commits you have

Okay, will do!

gdemonet committed:
Run Sphinx quickstart in docs/ directory

We leave the existing documents to be moved in a future commit.
We also rename the previous Makefile to a separate one to avoid
confusion. A later commit will include it in the main Makefile.

Issue: GH-1073

gdemonet committed:
Add Sphinx config overrides from MetalK8s 1.X

Mostly setting up the ReadTheDocs theme and a vendored Google Analytics
script (the pip package "sphinxcontrib-googleanalytics" is broken and
unmaintained).

Issue: GH-1073

gdemonet committed:
Add a 'livehtml' Make target

Allow users to watch the docs/ tree and build and serve HTML files from
the sources.

Issue: GH-1073

gdemonet committed:
Add a different index for LaTeX

Was used in MetalK8s 1.X to format the first pages of the PDF
documentation - we keep it only for the root index if tech writers need
to use the same pattern.

Issue: GH-1073

gdemonet committed:
Include a basic introduction

Copy-pasted from the MetalK8s 1.X docs.

Issue: GH-1073

gdemonet committed:
Ignore docs build artifacts

Issue: GH-1073

bert-e commented:
# History mismatch



Merge commit #c60e30afef0c5f8f850bbbbff3343a0fe4a4a4f0 on the integration branch
`w/2.1/feature/1073-bootstrap-sphinx-docs` is merging a branch which is neither the current
branch `feature/1073-bootstrap-sphinx-docs` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1073-bootstrap-sphinx-docs` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
@bert-e approve

bert-e commented:
# History mismatch



Merge commit #c60e30afef0c5f8f850bbbbff3343a0fe4a4a4f0 on the integration branch
`w/2.1/feature/1073-bootstrap-sphinx-docs` is merging a branch which is neither the current
branch `feature/1073-bootstrap-sphinx-docs` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1073-bootstrap-sphinx-docs` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

gdemonet commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1073-bootstrap-sphinx-docs` into
`development/2.0`
* [w/2.1/feature/1073-bootstrap-sphinx-docs](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/1073-bootstrap-sphinx-docs) will be merged into `development/2.1` 
* [w/2.2/feature/1073-bootstrap-sphinx-docs](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/1073-bootstrap-sphinx-docs) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye gdemonet.
